platforms/allwinner-d1/boards/lichee-rv/src/main.rs

```diff
@@ -0,0 +1,263 @@
+                }
+                Err(e) => {
+                    panic!("uhhhh {e:?}");
+                }
+            }
+        }
+    })
+    .unwrap();
+
+    // Loopback on virtual port zero
+    k.initialize(async move {
+        let mut hdl = loop {
+            match SerialMuxHandle::from_registry(k).await {
+                Some(c) => break c,
+                None => {
+                    k.sleep(Duration::from_millis(100)).await;
+                }
+            }
+        };
+
+        let hdl = hdl.open_port(0, 1024).await.unwrap();
+
+        loop {
+            let rx = hdl.consumer().read_grant().await;
+            let all_len = rx.len();
+            hdl.send(&rx).await;
+            rx.release(all_len);
+        }
+    })
+    .unwrap();
+
+    k.initialize(async move {
+        let mut hdl = loop {
+            match SerialMuxHandle::from_registry(k).await {
+                Some(c) => break c,
+                None => {
+                    k.sleep(Duration::from_millis(100)).await;
+                }
+            }
+        };
+
+        let hdl = hdl.open_port(1, 1024).await.unwrap();
+
+        loop {
+            hdl.send(b"Hello, world!\r\n").await;
+            k.sleep(Duration::from_secs(1)).await;
+        }
+    })
+    .unwrap();
+
+    timer0.set_prescaler(TimerPrescaler::P8); // 24M / 8:  3.00M ticks/s
+    timer1.set_prescaler(TimerPrescaler::P8);
+    timer0.set_mode(TimerMode::PERIODIC);
+    timer1.set_mode(TimerMode::SINGLE_COUNTING);
+    let _ = timer0.get_and_clear_interrupt();
+    let _ = timer1.get_and_clear_interrupt();
+
+    unsafe {
+        riscv::interrupt::enable();
+        riscv::register::mie::set_mext();
+    }
+
+    let plic = Plic::new(p.PLIC);
+
+    unsafe {
+        plic.register(Interrupt::TIMER1, timer1_int);
+        plic.register(Interrupt::DMAC_NS, handle_dmac);
+        plic.register(Interrupt::UART0, D1Uart::handle_uart0_int);
+
+        plic.activate(Interrupt::DMAC_NS, Priority::P1).unwrap();
+        plic.activate(Interrupt::UART0, Priority::P1).unwrap();
+    }
+
+    timer0.start_counter(0xFFFF_FFFF);
+
+    loop {
+        // Tick the scheduler
+        let start = timer0.current_value();
+        let tick = k.tick();
+
+        // Timer is downcounting
+        let elapsed = start.wrapping_sub(timer0.current_value());
+        let turn = k.timer().force_advance_ticks(elapsed.into());
+
+        // If there is nothing else scheduled, and we didn't just wake something up,
+        // sleep for some amount of time
+        if !tick.has_remaining && turn.expired != 0 {
+            let wfi_start = timer0.current_value();
+
+            // TODO(AJM): Sometimes there is no "next" in the timer wheel, even though there should
+            // be. Don't take lack of timer wheel presence as the ONLY heuristic of whether we
+            // should just wait for SOME interrupt to occur. For now, force a max sleep of 100ms
+            // which is still probably wrong.
```

Comment:
why is it probably wrong?

Reply:
Blindly waiting 100ms is not *great*, it keeps us from sleeping *forever*, which is something I saw during testing (it may have been because I wasn't previously checking the `expired` field of the timer wheel turn), but it also means that we could potentially sleep for longer than we should have.